File: 1145-binary-tree-coloring-game/1145-binary-tree-coloring-game.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} n
 * @param {number} x
 * @return {boolean}
 */
var btreeGameWinningMove = function(root, n, x) {
  let [leftCount, rightCount] = [0, 0];
  
  const count = (node) => {
    if (!node) return 0;
    
    const left = count(node.left);
    const right = count(node.right);
    if (node.val === x) [leftCount, rightCount] = [left, right];
    
    return left + right + 1;
  }
  
  count(root);
  
  return Math.max(leftCount, rightCount, n - 1 - leftCount - rightCount) > n / 2;
};

File: 1145-binary-tree-coloring-game/1145-binary-tree-coloring-game.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def btreeGameWinningMove(self, root: Optional[TreeNode], n: int, x: int) -> bool:
        left = right = 0
        
        def dfs(node):
            if not node:
                return 0
            l = dfs(node.left)
            r = dfs(node.right)
            if node.val == x:
                nonlocal left, right
                left, right = l, r
            return l + r + 1
          
          
        dfs(root)
        return max(left, right, n - 1 - left - right) * 2 > n